### PR TITLE
[test] add a line matcher object

### PR DESCRIPTION
**Pull request by picnixz**
Here's the line matching feature. 

TL;DR: Usage is 

```
app.stderr(flavor='re').assert_match('.*my message')
```

for the warnings for instance. It automatically removes the colors at the beginning and searches for a line matching the above string. The 'flavor' can maybe changed to 're' by default or have a `assert_re_match` method but that's how you would use the object in general.

There are some options with their default values but I'm not sure which one should be the default. For now, the line matcher object for test application uses the default options but maybe I can make it with better defaults. 

@chrisjsewell, I'd suggest you have a look at the tests (``tests/test_testing/test_matcher.py``) to see how it would be used. It's still WIP but I would like some feedback and what to expose (actually, with the current API you can more or less match whatever you want more or less however you want but it's better if you don't need to add options or if you have dedicated methods so as an extension's writer I'd be glad if you could give me some feedback on what would the functionalities that you want for sure). 

In the tests I used the most precise match information (namely the Line + offset) but you can ignore the offset and simply match against strings (see the tests for Line and Block objects).

I'm still wondering how to have a "nice" error context when the match fails... well, if you don't need the regex support you can just write something like ``app.stderr().lines() == [...]``. Maybe I can also expose a simple interface for simple matching.

### CODE REVIEW

**Comment by picnixz** [comment]
FYI: some stuff from opened PRs is included inside, so the diff will be smaller once they are merged. In addition, with the current implementation, it's easy for me to make some "shortcut" functions (the main logic is essentially done, except for the nice diff with regex patterns).

**Review by blowme8888 — APPROVED** [review]

**Comment by picnixz** [comment]
So, after some investigation, it's quite hard to have a "nice" diff because of regexes. I'll probably make another PR for that one because it's much more complex than what I thought at first (I think I'll take some inspiration from `difflib`, though it's not on my tasklist now).

I'll wait for some feedback before coming back to that PR.

File: sphinx/testing/_matcher/cleaner.py [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+def strip_chars(text: str, chars: StripChars = True, /) -> str:
+    """Strip expected characters from *text*."""
+    if isinstance(chars, bool):
+        return text.strip() if chars else text
+
+    if isinstance(chars, str) or chars is None:
+        return text.strip(chars)
+
+    msg = 'expecting a boolean, a string or None for %r, got: %r' % ('strip', chars)
+    raise ValueError(msg)
```
Comment (picnixz) [inline sphinx/testing/_matcher/cleaner.py]:
```suggestion
    return text.strip(chars)
```

I think I can let Python handle the exception itself.

File: sphinx/testing/_matcher/cleaner.py [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+def strip_lines(lines: Iterable[str], chars: StripChars = True, /) -> Iterable[str]:
+    """Call :meth:`str.strip` to each line in *lines*."""
+    if isinstance(chars, bool):
+        return map(str.strip, lines) if chars else lines
+
+    if isinstance(chars, str) or chars is None:
+        return (line.strip(chars) for line in lines)
+
+    msg = 'expecting a boolean, a string or None for %r, got: %r' % ('stripline', chars)
+    raise ValueError(msg)
```
Comment (picnixz) [inline sphinx/testing/_matcher/cleaner.py]:
```suggestion
    return (line.strip(chars) for line in lines)
```

Same here.

File: sphinx/testing/_matcher/cleaner.py [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+    empty = get_option(options, 'empty')
+    compress = get_option(options, 'compress')
+    unique = get_option(options, 'unique')
+    lines = filter_lines(lines, empty=empty, compress=compress, unique=unique)
+
+    delete = get_option(options, 'delete')
+    flavor = get_option(options, 'flavor')
+    lines = prune(lines, delete, flavor=flavor)
+
+    return ignore_lines(lines, get_option(options, 'ignore'))
```
Comment (picnixz) [inline sphinx/testing/_matcher/cleaner.py]:
```suggestion
    lines = prune(lines, delete, flavor=flavor)

    ignore = get_option(options, 'ignore')
    lines = ignore_lines(lines, ignore)

    return lines
```

Maybe for clarity.

File: sphinx/testing/_matcher/buffer.py [hunk 4]
```diff
@@ -0,0 +1,444 @@
...
+    __slots__ = ('__buffer', '__offset', '__weakref__')
+
+    def __init__(self, buffer: _T, offset: int = 0, /, *, _check: bool = True) -> None:
+        if _check and not isinstance(offset, int):
+            msg = f'offset must be an integer, got: {offset!r}'
+            raise TypeError(msg)
+
+        if _check and offset < 0:
+            msg = f'offset must be >= 0, got: {offset!r}'
+            raise ValueError(msg)
```
Comment (picnixz) [inline sphinx/testing/_matcher/buffer.py]:
```suggestion
        if _check:
            if not isinstance(offset, int):
                msg = f'offset must be an integer, got: {offset!r}'
                raise TypeError(msg)

            if offset < 0:
                msg = f'offset must be >= 0, got: {offset!r}'
                raise ValueError(msg)
```

File: sphinx/testing/_matcher/buffer.py [hunk 5]
```diff
@@ -0,0 +1,444 @@
...
+    def index(self, sub: str, start: int = 0, end: int | None = None, /) -> int:
+        """See :meth:`str.index`."""
+        return self.buffer.index(sub, start, end)
+
+    def rindex(self, sub: str, start: int = 0, end: int | None = None, /) -> int:
+        """See :meth:`str.rindex`."""
+        return self.buffer.rindex(sub, start, end)
+
+    def __cast(self, other: object, /) -> Self | object:
+        """Try to parse *object* as a :class:`Line`."""
```
Comment (picnixz) [inline sphinx/testing/_matcher/buffer.py]:
```suggestion
        """Try to parse *other* as a :class:`Line`."""
```

File: sphinx/testing/_matcher/buffer.py [hunk 6]
```diff
@@ -0,0 +1,444 @@
...
+        if isinstance(other, self.__class__):
+            return other
+        if isinstance(other, str):
+            return self.__class__(other, self.offset, _check=False)
```
Comment (picnixz) [inline sphinx/testing/_matcher/buffer.py]:
```suggestion
            return self.__class__(other, self.offset, _check=True)
```

File: sphinx/testing/_matcher/buffer.py [hunk 7]
```diff
@@ -0,0 +1,444 @@
...
+        :param limit: The number of lines in the source the block belongs to.
+        :return: The slices for the 'before' and 'after' lines.
+
+        Example::
+
+           source = ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10']
+           block = Block(['4', '5', '6'], 3)
+           before, after = block.context(2, 10)
+           assert source[before] == ['2', '3']
+           assert source[after] == ['7', '8']
```
Comment (picnixz) [inline sphinx/testing/_matcher/buffer.py]:
```suggestion
        Example::

            source = ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10']
            block = Block(['4', '5', '6'], 3)
            before, after = block.context(2, 10)
            assert source[before] == ['2', '3']
            assert source[after] == ['7', '8']
```

File: sphinx/testing/_matcher/buffer.py [hunk 8]
```diff
@@ -0,0 +1,444 @@
...
+        value: object,
+        start: SupportsIndex = 0,
+        stop: SupportsIndex = sys.maxsize,
+        /,
+    ) -> int:
+        """See :meth:`tuple.count`."""
+        return self.buffer.index(value, start, stop)
+
+    def count(self, value: object, /) -> int:
+        """See :meth:`tuple.count`."""
```
Comment (picnixz) [inline sphinx/testing/_matcher/buffer.py:417]:
```suggestion
        """See :meth:`tuple.index`."""
```

File: sphinx/testing/_matcher/buffer.py [hunk 9]
```diff
@@ -0,0 +1,444 @@
...
+        return self.buffer.count(value)
+
+    def __cast(
+        self, other: Sequence[object] | tuple[Sequence[object], int], /
+    ) -> tuple[Sequence[object], int]:
+        """Try to parse *object* as a pair ``(lines, block offset)``."""
```
Comment (picnixz) [inline sphinx/testing/_matcher/buffer.py]:
```suggestion
        """Try to parse *other* as a pair ``(lines, block offset)``."""
```

File: sphinx/testing/_matcher/engine.py [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+        to_line_patterns("my pattern") == to_line_patterns(["my pattern"])
+    """
+    if isinstance(expect, (str, re.Pattern)):
+        return (expect,)
+
+    def key(x: str | re.Pattern[str]) -> str:
+        return x if isinstance(x, str) else x.pattern
+
+    if optimized:
+        return sorted(set(expect), key=key)
```
Comment (picnixz) [inline sphinx/testing/_matcher/engine.py]:
```suggestion
        return tuple(sorted(set(expect), key=key))
```

File: sphinx/testing/_matcher/engine.py [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+
+# fmt: off
+@overload
+def transform(fn: Callable[[str], str], x: str) -> str: ...  # NoQA: E704
+@overload
+def transform(fn: Callable[[str], str], x: re.Pattern[str]) -> re.Pattern[str]: ...  # NoQA: E704
+# fmt: on
+def transform(fn: Callable[[str], str], x: LinePattern) -> LinePattern:  # NoQA: E302
+    """Transform regular expressions, leaving compiled patterns untouched."""
+    return fn(x) if isinstance(x, str) else x
```
Comment (picnixz) [inline sphinx/testing/_matcher/engine.py]:
```suggestion
# fmt: off
@overload
def transform(fn: Callable[[str], str], x: str, /) -> str: ...  # NoQA: E704
@overload
def transform(fn: Callable[[str], str], x: re.Pattern[str], /) -> re.Pattern[str]: ...  # NoQA: E704
# fmt: on
def transform(fn: Callable[[str], str], x: LinePattern, /) -> LinePattern:  # NoQA: E302
    """Transform regular expressions, leaving compiled patterns untouched."""
    return fn(x) if isinstance(x, str) else x
```

File: sphinx/testing/_matcher/options.py [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+    of the given strings or matches any of the given patterns.
+
+    This transformation is applied at the end of the transformation
+    chain, just before filtering the output lines are filtered with
+    the :attr:`ignore` predicate.
+
+    Example::
+
+        clean('abcdA\n1', delete='abcd') == ['A', '1']
+        clean('1234A\nxyzt', delete=r'\d+', flavor='re') == ['A', 'xyzt']
```
Comment (picnixz) [inline sphinx/testing/_matcher/options.py]: This is now outdated (I'll add a ``.. seealso::``)

```suggestion
```

File: sphinx/testing/_matcher/options.py [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+    compress=False,
+    unique=False,
+    delete=(),
+    ignore=None,
+    flavor='none',
+)
+"""The default (read-only) options values."""
+
+if TYPE_CHECKING:
+    _OptionsView = Union[Options, CompleteOptions]
```
Comment (picnixz) [inline sphinx/testing/_matcher/options.py]: I'm putting it here to avoid have explicit forward references at the top of the file.

File: sphinx/testing/_matcher/options.py [hunk 14]
```diff
@@ -0,0 +1,241 @@
...
+
+
+# Disable the ruff formatter to minimize the number of empty lines.
+#
```
Comment (picnixz) [inline sphinx/testing/_matcher/options.py]:
```suggestion
```

File: sphinx/testing/_matcher/util.py [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+    # use the C API to efficiently consume iterators
+    if n is None:
+        deque(iterator, maxlen=0)
+    else:
+        n = max(n, 0)
+        next(itertools.islice(iterator, n, n), None)
+
+
+def unique_justseen(iterable: Iterable[_T], /) -> Iterator[_T]:
+    """Yields elements in order, ignoring serial duplicates.
```
Comment (picnixz) [inline sphinx/testing/_matcher/util.py]:
```suggestion
    """Yield elements in order, ignoring serial duplicates.
```

File: sphinx/testing/_matcher/util.py [hunk 16]
```diff
@@ -0,0 +1,216 @@
...
+
+    Credits go to :func:`!more_itertools.recipes.unique_justseen`.
+    """
+    return map(next, map(itemgetter(1), itertools.groupby(iterable)))
+
+
+def unique_everseen(iterable: Iterable[_T], /) -> Iterator[_T]:
+    """Yields elements in order, ignoring duplicates.
```
Comment (picnixz) [inline sphinx/testing/_matcher/util.py]:
```suggestion
    """Yield elements in order, ignoring duplicates.
```

File: sphinx/testing/matcher.py [hunk 17]
```diff
@@ -0,0 +1,329 @@
...
+        count: int | None = None,
+        flavor: Flavor | None = None,
+    ) -> None:
+        """Assert that there exist one or more lines matching *pattern*.
+
+        :param expect: One or more patterns the lines must satisfy.
+        :param count: If specified, the exact number of matching lines.
+        :param flavor: Optional temporary flavor for string patterns.
+        """
+        patterns = engine.to_line_patterns(expect)
```
Comment (picnixz) [inline sphinx/testing/matcher.py]:
```suggestion
        patterns = engine.to_line_patterns(expect, optimized=True)
```

File: sphinx/testing/matcher.py [hunk 18]
```diff
@@ -0,0 +1,329 @@
...
+        context: int = 3,
+        flavor: Flavor | None = None,
+    ) -> None:
+        """Assert that there are no lines matching *pattern*.
+
+        :param expect: One or more patterns the lines must not satisfy.
+        :param context: Number of lines to print around a failing line.
+        :param flavor: Optional temporary flavor for string patterns.
+        """
+        patterns = engine.to_line_patterns(expect)
```
Comment (picnixz) [inline sphinx/testing/matcher.py]:
```suggestion
        patterns = engine.to_line_patterns(expect, optimized=True)
```

File: sphinx/testing/matcher.py [hunk 19]
```diff
@@ -0,0 +1,329 @@
...
+           This interface does not support single :class:`~re.Pattern`
+           objects as they could be interpreted as a line or a block
+           pattern.
+        """
+        patterns = engine.to_block_pattern(expect)
+
+        lines = self.lines()
+        # early abort if there are more expected lines than actual ones
+        if (width := len(patterns)) > len(lines):
+            return
```
Comment (picnixz) [inline sphinx/testing/matcher.py]: I did not consider the case ``width = 0``. So I'll need to fix that tomorrow.

File: sphinx/testing/matcher.py [hunk 20]
```diff
@@ -0,0 +1,329 @@
...
+    def _assert_not_found(
+        self,
+        pattern_type: PatternType,
+        patterns: Sequence[LinePattern],
+        *,
+        context_size: int,
+        flavor: Flavor | None = None,
+    ) -> None:
+        lines = self.lines()
+        if (count := len(patterns)) > len(lines):
```
Comment (picnixz) [inline sphinx/testing/matcher.py]: Same here.

**Review by picnixz — COMMENTED** [review]
Some comments for myself.

EDIT: found a bug in the `Block.__lt__` and `Block.__gt__` methods.

File: sphinx/testing/_matcher/buffer.py [hunk 21]
```diff
@@ -0,0 +1,444 @@
...
+    # exposed :class:`tuple` interface
+
+    def index(
+        self,
+        value: object,
+        start: SupportsIndex = 0,
+        stop: SupportsIndex = sys.maxsize,
+        /,
+    ) -> int:
+        """See :meth:`tuple.count`."""
```
Comment (picnixz) [inline sphinx/testing/_matcher/buffer.py]:
```suggestion
        """See :meth:`tuple.index`."""
```

**Comment by picnixz** [comment]
After considering #12216, I'll update this PR as well.

**Comment by picnixz** [comment]
@jayaddison @chrisjsewell When you have time, could you have a look at this one? I know that there are a lot files and a lot of changes, but I don't mind you to take your time (I'll be away for a few weeks). I'd suggest you to open the files in an IDE instead on github (especially the test files).

When I'm back, I'll write some separate docs with more examples but it'd be good if I can have some feedback. I've implemented some cleaning functionalities that are all disabled by default when the matcher is used as a standalone object but some capabilities are enabled for the testing plugin (see `options.py` for the supported options and `cleaner.py` for the corresponding implementation). 

You can essentially do:

```
app.stderr().lines() == ['my first line', 'my second line']
```

without bother about colors or

```
app.stderr(flavor='re').assert_block(['^.+/index.rst: my first line but with regex', '^.+/index.rst: my second line but with regex'])
```

You can also assert that a line exists/does not exist, the number of occurrences of that line. You can also extract the lines / blocks matched by certain patterns if you need a more complex comparison / post-processing step.

There are other variants where you can remove the path prefix (and I can do add some public method such as `assert_block_without_path()` that would the job for you) by relying on the current public API. Since I wanted to avoid having a slow implementation, I tried to rely as much as possible on the C API and since I wanted to protect against mutable lines, I also chose tuples instead of lists whenever possible (it's also more efficient since we don't need to re-allocate everything). 

Note that you can still access the raw content by using `app.status` or `app.warning` and you can also use the cleaner functions yourself directly on strings without relying on the matcher's API if you're not happy.

The `_util.py` module contains some utility functions that are essentially used for debugging your output and itertools recipes. As for the `_engine.py` module, it is responsible to handle regular expressions and how to translate them (you can have "exact" match, "regex" match or "fnmatch" match, but I think the last one wouldn't be used much, but still, it's supported accordingly). 

For such implementation, the `test_matcher_buffer.py` and `test_matcher.py` are a good starting point where you can see how the high-level API is used and how the low-level API is implemented (and what you can).

**Comment by chrisjsewell** [comment]
> When you have time, could you have a look at this one? 
> but I don't mind you to take your time (I'll be away for a few weeks).

Thanks @picnixz I'll definitely try 😅 but yes as you are away (thanks for letting me know) I will put it a little lower on the TODO list.
When you get back, feel free to ping me if I haven't yet!

**Comment by picnixz** [comment]
Sure! I just pushed some docs because there multiple typos and refactored a bit things. By the way, I also observed that we had a bug in autodoc thanks to that... so I fixed it (I've got 3 PRs that need to be merged before that just for the docs to be correctly rendered).

**Comment by jayaddison** [comment]
> TL;DR: Usage is
> 
> ```
> app.stderr(flavor='re').assert_match('.*my message')
> ```

Checking some of the design reasoning for the method signature:

  * A reason to prefer the caller to select a `flavor` is that otherwise, a matcher pattern of type `str` could ambiguously be either a regex or a simple string match (`app.stderr.assert_match('A*')`).
  * Using keyword-arguments to select the flavor (`app.stderr.assert_match(string='[not a regex]')` might be non-obvious, and would prevent positional single-argument usage like the example above.

File: sphinx/testing/matcher/__init__.py [hunk 22]
```diff
@@ -0,0 +1,365 @@
...
+        By convention, the following are equivalent::
+
+            matcher.iterfind('line to find', ...)
+            matcher.iterfind(['line to find'], ...)
+        """
+        patterns = _engine.to_line_patterns(patterns)
+        if not patterns:  # nothinig to match
+            return
+
+        compiled_patterns = set(self.__compile(patterns, flavor=flavor))
```
Comment (jayaddison) [inline sphinx/testing/matcher/__init__.py]: What's the benefit in de-duplicating the compiled patterns?

File: tests/test_testing/test_matcher_engine.py [hunk 23]
```diff
@@ -0,0 +1,87 @@
...
+    # ensure build reproducibility
+    assert engine.to_line_patterns({'a', p}) == ('a', p)
+
+    p1 = re.compile('a')
+    p2 = re.compile('a', re.MULTILINE)
+    p3 = re.compile('ab')
+    p4 = re.compile('ab', re.MULTILINE)
+    ps = (p1, p2, p3, p4)
+
+    for _ in range(100):
```
Comment (jayaddison) [inline tests/test_testing/test_matcher_engine.py:28]: Do we anticipate that additional cases are likely to be added here?  (`4! = 24`, so if not, perhaps we should use `itertools` to generate the permutations instead)

File: sphinx/testing/matcher/__init__.py (hunk 22 above)
Comment (picnixz) [reply sphinx/testing/matcher/__init__.py]: Efficiency, just in case. Since it's likely to be used in some testing code, I predict that there might be copy-paste here and there. But technically, I can ignore that and just say "you should know what to do".

File: tests/test_testing/test_matcher_engine.py (hunk 23 above)
Comment (picnixz) [reply tests/test_testing/test_matcher_engine.py:28]: Err.... Actually no. Good catch ! (I think it's something I forgot to change).

**Comment by picnixz** [comment]
> A reason to prefer the caller to select a flavor is that otherwise, a matcher pattern of type str could ambiguously be either a regex or a simple string match (app.stderr.assert_match('A*')).

I don't know which flavor to have by default... I would like to say "yay, it's maybe better to have a pure string flavor" because you usually want to have an exact match (I think most of our tests have exact matching and I think people don't want to escape possible meta-characters...).

One possibility that I had in mind is just to have other methods:

- `assert_match`
- `assert_equal`

But I'm not very happy with the `equal` itself... (ideally, it's `assert_that_there_is_a_line_equal_to_one_of_the_strings_or_patterns`)

> Using keyword-arguments to select the flavor (app.stderr.assert_match(string='[not a regex]') might be non-obvious, and would prevent positional single-argument usage like the example above.

I think `assert_match(expect, flavor=...)` is more natural but I'm open to suggestions. Alternatively, we could have `assert_match(regex='whatever to be considered a regex', string='whatever string it should be', fnmatch='whatever fnmatch pattern')` but I'm not sure if more flavors should be supported (I tried to have an interface that is flavor-agnostic as much as possible and delegate the task to converting string-like objects to re-patterns objects however you deem them fit (maybe I could even use predicate-based things because I don't think I use anything else except `pattern.match`)).

File: sphinx/testing/matcher/__init__.py (hunk 22 above)
Comment (jayaddison) [reply sphinx/testing/matcher/__init__.py]: Ok, I agree with the efficiency argument, and would also note that de-duplicating regexes shouldn't ever cause change-in-behaviour - if `x` matches `y`, it'll always match, whether it's checked once or ten times -- so let's check once.

But re-ordering the order in which matches are applied _could_ be hard-to-debug; so I think I'd prefer an order-preserving dedup here if that's OK with you?

(to explain: a human is likely to order regexes based on how likely they think they are to catch problems -- and it's possible that they could re-arrange those to be performance-optimized, whereas a Python `set` could disrupt that ordering)

**Comment by jayaddison** [comment]
> I think `assert_match(expect, flavor=...)` is more natural but I'm open to suggestions.

I think I'd probably prefer a developer experience of:

  * `assert_contains(string)` to check for lines containing `string`.
  * `assert_matches(pattern)` to check for lines that match a regular expression `pattern`.

That is: binding the flavour to the method name.  That's partly because it could help me to think about and write / read code, but also there are some second-order downstream static analysis/dependency benefits (tooling/analysts can infer to some degree that `assert_contains` does not enter finite automata logic originated from Sphinx, for example).

File: sphinx/testing/matcher/__init__.py (hunk 22 above)
Comment (picnixz) [reply sphinx/testing/matcher/__init__.py]: Ah, fine by me. I have a utility function for that (it's the unique_everseen one)

**Comment by AA-Turner** [comment]
I am probably missing some context here -- can someone point me to a brief overview of what this PR is for? The body text of the description simply says here is the matcher!

I'm aware Benedict has his thesis coming up soon, so no rush, just would be interested in the background given it's a 3000 line PR.

A

**Comment by picnixz** [comment]
> I am probably missing some context here 

Oh yes, I think I never created an issue for that and just rushed to the implementation as if it were a hackathon. So, here's the context: we were doing some cleanups about colorization and co and I observed that many tests always have the same logic, namely:

```lines = strip_colors(app.status.getvalue()).strip().splitlines()``` 

and then, they match the lines one by one. All those lines could be replaced by a single function, let's say `app.get_status_lines()` and that would be it. But since I had time on my hands (and I wanted to have something for my local dev as well), I ended up implementing a line matcher where you can do more than just matching strings by strings, but also match blocks in one go, or with regexes and so on + have a nicer diff. The original idea was based on the pytester object that I used for testing plugins.

Technically speaking, the original issue could be solved by just adding two methods to `SphinxTestApp` that would just give you the lines. I can do a small PR if you think the 3k lines are too much. Actually, the interface of the matcher object is flexible enough that we can gradually add more methods if needed, but I essentially built it so that the most common operations are done by default, namely "remove colors -> strip -> split lines without keeping line breaks" (well, this default sequence is only for the matcher object for the test application, because people might want to use that matcher for any other kind of string technically).

I also think that it could be useful for matching autodoc outputs. I created a class for formatting expected RST output for enumeration test cases and I thought "how nice it would be to be able to do that but for other autodoc outputs as well" so I think the matcher object can be used in conjunction with those factory classes in a more user-friendly API. At least, tests would be easier to write (and probably shorter). Also, while the pytest output is fine, sometimes it's hard to detect exactly where the error happenned in the huge diff (the line matcher object would "highlight" the erroneous blocks; for now the logic is simple enough but I intended to do it using diff-like algorithms).

> I'm aware Bénédikt has his thesis coming up soon, so no rush, just would be interested in the background given it's a 3000 line PR.

Yes, I'm currently writing it !

**Comment by picnixz** [comment]
@jayaddison 

> That is: binding the flavour to the method name. That's partly because it could help me to think about and write / read code, but also there are some second-order downstream static analysis/dependency benefits (tooling/analysts can infer to some degree that assert_contains does not enter finite automata logic originated from Sphinx, for example).

The argument is sound. I can make the flavor-agnostic method private and expose dispatcher methods.

File: tests/test_testing/test_matcher_engine.py (hunk 23 above)
Comment (picnixz) [reply tests/test_testing/test_matcher_engine.py:28]: Ah ! I remember why I did this ! it's to be sure that sets are considered correctly (I mean, the type `set`). The reason is that I wanted to have multiple objects of the type `set` that are independent to each other.
